File: ggcommons/messaging/providers/greengrass/iotcore_subscription_handler.py
import json
import logging
from typing import Callable
from ggcommons.messaging.message import Message
from ggcommons.messaging.providers.greengrass.subscription_handler import (
    SubscriptionHandler,
)

logger = logging.getLogger("ConfigManager")


class IotCoreSubscriptionHandler(SubscriptionHandler):
    def __init__(
        self,
        topic_filter,
        callback: Callable[[str, Message], None],
        serialize_processing: bool = False,
    ):
        super().__init__(topic_filter, callback, serialize_processing)

    def parse_raw_payload(self, event) -> (str, dict):
        received_payload = json.loads(str(event.message.payload, "utf-8"))
        logger.debug(
            f"IoT Core: common: PubSubDataHandler: on_stream_event: subscribed message: {received_payload}"
        )
        return event.message.topic_name, received_payload

File: src/main/java/com/aws/proseve/ggcommons/messaging/providers/greengrass/IotCoreSubscriptionHandler.java
package com.aws.proseve.ggcommons.messaging.providers.greengrass;


import com.github.cliftonlabs.json_simple.Jsoner;
import com.aws.proseve.ggcommons.messaging.Message;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import oshi.util.tuples.Pair;
import software.amazon.awssdk.aws.greengrass.model.IoTCoreMessage;
import software.amazon.awssdk.aws.greengrass.model.MQTTMessage;

import java.nio.charset.StandardCharsets;
import java.util.function.BiConsumer;

public class IotCoreSubscriptionHandler extends SubscriptionHandler<IoTCoreMessage>
{
    protected static final Logger LOGGER = LogManager.getLogger(IotCoreSubscriptionHandler.class);

    public IotCoreSubscriptionHandler(String topicFilter, BiConsumer<String, Message> callback, int maxConcurrency)
    {
        super(topicFilter, callback, maxConcurrency);
    }

    @Override
    Pair<String, Message> parseRawPayload(IoTCoreMessage iotCoreMessage)
    {
        LOGGER.debug("IoT Core message received on subscription to topic filter '{}'", topicFilter);
        Pair<String, Message> retVal = null;
        try
        {
            String topic = iotCoreMessage.getMessage().getTopicName();
            MQTTMessage mqttMessage = iotCoreMessage.getMessage();
            String msgChars = new String(mqttMessage.getPayload(), StandardCharsets.UTF_8);
            Message msg;
            try
            {
                msg = Message.build(Jsoner.deserialize(msgChars));
            }
            catch (Exception e)
            {
                msg = Message.build(msgChars);
            }
            retVal = new Pair<>(topic, msg);
        } catch (Exception e) {
            LOGGER.warn("Problem decoding IoT Core payload into Message on topic {}: {}.  Ignoring message.",
                    topicFilter, e.toString());
        }
        return retVal;
    }
}
